fontforge/splinestroke.c

```diff
@@ -32,3689 +32,1805 @@
-	if ( cnt<4 ) cnt=4;
-	if ( cnt&1 )
-	    ++cnt;
-	if ( isend ) {
-	    start=2*cnt; end=0; incr=-1;
-	} else {
-	    start=0; end=2*cnt; incr=1;
-	    slope.x = -slope.x;
-	    slope.y = -slope.y;
-	}
-	base2.x = base.x + (c->radius) * slope.x;
-	base2.y = base.y + (c->radius) * slope.y;
-	halfleft.x = base2.x - c->radius*done.slope.y;
-	halfleft.y = base2.y + c->radius*done.slope.x;
-	halfright.x = base2.x + c->radius*done.slope.y;
-	halfright.y = base2.y - c->radius*done.slope.x;
-	{
-	    struct extrapoly *ep;
-	    if ( c->ecur>=c->emax )
-		c->ep = realloc(c->ep,(c->emax+=40)*sizeof(struct extrapoly));
-	    ep = &c->ep[c->ecur++];
-	    ep->poly[0] = done.left;
-	    ep->poly[1].x = done.left.x + c->radius*slope.x;
-	    ep->poly[1].y = done.left.y + c->radius*slope.y;
-	    ep->poly[2].x = done.right.x + c->radius*slope.x;
-	    ep->poly[2].y = done.right.y + c->radius*slope.y;
-	    ep->poly[3] = done.right;
-	    if ( !isend ) {
-		BasePoint temp;
-		ep->poly[0] = done.right;
-		temp = ep->poly[1];
-		ep->poly[1] = ep->poly[2];
-		ep->poly[2] = temp;
-		ep->poly[3] = done.left;
-	    }
-	    ep->ptcnt = 4;
-	}
-	for ( i=start; ; i+=incr ) {
-	    p = &c->all[c->cur++];
-	    p->sp = done.sp;
-	    p->t = isend;
-	    p->line = true;
-	    p->me = base;
-	    p->slope = slope;
-	    p->needs_point_left = p->needs_point_right = (i==0 || i==cnt || i==cnt+cnt);
-	    if ( i<=cnt ) {
-		factor = c->radius*i/cnt;
-		p->left.x = base2.x - factor*done.slope.y;
-		p->left.y = base2.y + factor*done.slope.x;
-		p->right.x = base2.x + factor*done.slope.y;
-		p->right.y = base2.y - factor*done.slope.x;
-	    } else {
-		factor = c->radius*(i-cnt)/cnt;
-		p->left.x = halfleft.x - factor*slope.x;
-		p->left.y = halfleft.y - factor*slope.y;
-		p->right.x = halfright.x - factor*slope.x;
-		p->right.y = halfright.y - factor*slope.y;
-	    }
-	    if ( i==end )
-	break;
-	}
-    } else {
-	/* Semicircle */
-	if ( cnt<8 )
-	    cnt = 8;
-	if ( isend ) {
-	    start=cnt; end=0; incr=-1;
-	} else {
-	    start=0; end=cnt; incr=1;
-	}
-	for ( i=start; ; i+=incr ) {
-	    p = &c->all[c->cur++];
-	    *p = done;
-	    p->circle = true;
-	    p->needs_point_left = p->needs_point_right = i==0 || i==cnt;
-	    si = sin((PI/2.0)*i/(bigreal) cnt);
-	    co = sqrt(1-si*si);
-	    if ( isend ) co=-co;
-	    slope2.x = slope.x*co + slope.y*si;
-	    slope2.y =-slope.x*si + slope.y*co;
-	    p->left.x = base.x - c->radius*slope2.x;
-	    p->left.y = base.y - c->radius*slope2.y;
-	    slope2.x = slope.x*co - slope.y*si;
-	    slope2.y = slope.x*si + slope.y*co;
-	    p->right.x = base.x - c->radius*slope2.x;
-	    p->right.y = base.y - c->radius*slope2.y;
-	    if ( i==end )
-	break;
-	}
-    }
-    if ( !isend )
-	c->all[c->cur++] = done;
+StrokeInfo *InitializeStrokeInfo(StrokeInfo *sip) {
+    if ( sip==NULL )
+	sip = malloc(sizeof(StrokeInfo));
+
+    *sip = (StrokeInfo) {
+        25.0,         // radius
```

Comment:
Excuse my ignorance, but how do these get determined?

Reply:
There are five ways:

1. The radius default mostly serves as a CVStroke default. The other interfaces make you set it in cases where it will be used (i.e. pen types other than si_nib). 
2. The pen, cap and join defaults represent offsetting with round ends. This seems like a good general default behavior. 
3. The rest of the boolean, 0.0, and NULL values are "principled" defaults: there are options and these are the best initial choices. I actually added this function because I got tired of dealing with boolean conditions "phrased" so that false was default to work with memset(0).
4. A join limit of 20 corresponds to an SVG `stroke-miterlimit` of 20 and therefore a PostScript miterlimit of 10 (PS used stroke width, SVG uses offsetting radius). That's from the `c.miterlimit` assignment in the old `SplineSetStroke()`, as noted in the old comments. 
5. The accuracy target is my choice. I'm not sure I can defend it on principles so I'm open to suggestion and argument.

Reply:
Actually, with the lc_nib and lj_nib defaults it's arguably less relevant that the circular nib will have round caps and joins as that these are the appropriate defaults for all other nib types, given that the previous algorithm only had these options for rectangular and polygonal nibs. These are the options that correspond to what would happen if the nib were traced on paper.